attest/ake/src/initiator.rs

```diff
@@ -212,19 +210,90 @@ where
         self,
         _csprng: &mut R,
         input: UnverifiedAttestationEvidence,
-    ) -> Result<(Terminated, VerificationReport), Self::Error> {
+    ) -> Result<(Terminated, EvidenceKind), Self::Error> {
         let output = self
             .state
             .read_message(input.as_ref())
             .map_err(Error::HandshakeRead)?;
         match output.status {
             HandshakeStatus::InProgress(_state) => Err(Error::HandshakeNotComplete),
             HandshakeStatus::Complete(_) => {
-                let remote_report = VerificationReport::decode(output.payload.as_slice())
-                    .map_err(|_e| Error::AttestationEvidenceDeserialization)?;
+                if let Ok(remote_evidence) = EvidenceKind::from_bytes(output.payload.as_slice()) {
+                    Ok((Terminated, remote_evidence))
+                } else {
+                    let remote_report = VerificationReport::decode(output.payload.as_slice())
+                        .map_err(|_| Error::AttestationEvidenceDeserialization)?;
 
-                Ok((Terminated, remote_report))
+                    Ok((Terminated, EvidenceKind::Epid(remote_report)))
+                }
             }
         }
     }
 }
+
+fn verify_evidence_kind<PubKey: KexPublic>(
+    attestation_evidence: EvidenceKind,
+    identities: &Vec<TrustedIdentity>,
+    time: Option<DateTime>,
+    remote_identity: &PubKey,
+) -> Result<(), Error> {
+    match attestation_evidence {
+        EvidenceKind::Dcap(dcap_evidence) => {
+            let dcap_evidence: DcapEvidence = (&dcap_evidence)
+                .try_into()
+                .map_err(|_| Error::AttestationEvidenceDeserialization)?;
+            verify_dcap_evidence(dcap_evidence, identities, time)
+        }
+        EvidenceKind::Epid(verification_report) => {
+            verify_verification_report(&verification_report, identities, remote_identity)
+        }
+    }
+}
+
+fn verify_dcap_evidence(
+    dcap_evidence: DcapEvidence,
+    identities: &Vec<TrustedIdentity>,
+    time: Option<DateTime>,
+) -> Result<(), Error> {
+    let DcapEvidence {
+        quote,
+        collateral,
+        report_data,
+    } = dcap_evidence;
+    let (quote, collateral, report_data) = (quote, collateral, report_data);
+
+    let verifier = DcapVerifier::new(identities, time, report_data);
```

Comment:
This should be ensuring that the `report_data.key` is the same as the `remote_identity`

```suggestion
    if remote_identity.map_bytes(|bytes| bytes != report_data.key().to_bytes().as_slice()) {
        return Err(Error::BadRemoteIdentity);
    }

    let verifier = DcapVerifier::new(identities, time, report_data);
```